File: bindings/typescript/src/kernel.ts
import { existsSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

export const KERNEL_SCHEMA_VERSION = "1.0" as const;

export type KernelOperation =
  | "discover_models"
  | "fit_model"
  | "predict_model"
  | "simulate_model"
  | "summarize_model"
  | "diagnose_model";

export interface KernelRequestPayload {
  [key: string]: unknown;
}

export interface KernelRequestMetadata {
  [key: string]: unknown;
}

export interface KernelRequest {
  schema_version: string;
  operation: KernelOperation;
  model_key: string | null;
  payload: KernelRequestPayload;
  metadata: KernelRequestMetadata;
}

function hasKernelRoot(candidate: string): boolean {
  return (
    existsSync(join(candidate, "src", "innovate", "kernel.py")) &&
    existsSync(join(candidate, "conductor", "tracks.md"))
  );
}

export function kernelRepoRoot(startDir = dirname(fileURLToPath(import.meta.url))): string {
  if (process.env.INNOVATE_REPO_ROOT) {
    return resolve(process.env.INNOVATE_REPO_ROOT);
  }

  let current = resolve(startDir);
  while (true) {
    if (hasKernelRoot(current)) {
      return current;
    }

    const parent = dirname(current);
    if (parent === current) {
      throw new Error("Unable to locate the innovate repository root");
    }

    current = parent;
  }
}

export function kernelRepoRootOrNull(startDir = dirname(fileURLToPath(import.meta.url))): string | null {
  try {
    return kernelRepoRoot(startDir);
  } catch {
    return null;
  }
}

export function kernelPackageRoot(): string {
  const moduleDir = dirname(fileURLToPath(import.meta.url));
  return moduleDir.endsWith("dist") ? dirname(moduleDir) : dirname(moduleDir);
}

export function kernelBindingsRoot(): string {
  const repoRoot = kernelRepoRootOrNull();
  return repoRoot ? join(repoRoot, "bindings", "typescript") : kernelPackageRoot();
}

export function kernelBridgeScript(): string {
  const repoBridge = join(kernelBindingsRoot(), "inst", "python", "kernel_bridge.py");
  if (existsSync(repoBridge)) {
    return repoBridge;
  }

  const packagedBridge = join(kernelPackageRoot(), "inst", "python", "kernel_bridge.py");
  if (existsSync(packagedBridge)) {
    return packagedBridge;
  }

  throw new Error("Unable to locate the packaged Innovate kernel bridge");
}

export function kernelPythonCommand(): string {
  return process.env.INNOVATE_PYTHON_COMMAND || "uv run python";
}

export function kernelSchemaVersion(): string {
  return KERNEL_SCHEMA_VERSION;
}

export function kernelRequest(
  operation: KernelOperation,
  modelKey: string | null = null,
  payload: KernelRequestPayload = {},
  metadata: KernelRequestMetadata = {},
  schemaVersion: string = KERNEL_SCHEMA_VERSION,
): KernelRequest {
  if (!operation) {
    throw new Error("kernelRequest() requires a non-empty operation");
  }

  if (operation !== "discover_models" && !modelKey) {
    throw new Error(`Kernel operation '${operation}' requires a model_key`);
  }

  return {
    schema_version: schemaVersion,
    operation,
    model_key: modelKey,
    payload,
    metadata,
  };
}

File: bindings/go/kernel.go
package innovate

import (
	"path/filepath"
	"runtime"
)

const kernelSchemaVersionValue = "1.0"

// ModuleRoot returns the Go bindings module root on disk.
func ModuleRoot() string {
	return moduleRoot()
}

// BridgeScriptPath returns the Python bridge entrypoint used by the Go layer.
func BridgeScriptPath() string {
	return filepath.Join(moduleRoot(), "inst", "python", "kernel_bridge.py")
}

// KernelSchemaVersion returns the shared kernel contract schema version.
func KernelSchemaVersion() string {
	return kernelSchemaVersionValue
}

// KernelOperations returns the stable kernel operations exposed by the Go bindings.
func KernelOperations() []KernelOperation {
	return []KernelOperation{
		KernelOperation(kernelOperationDiscoverModels),
		KernelOperation(kernelOperationFitModel),
		KernelOperation(kernelOperationPredictModel),
		KernelOperation(kernelOperationSimulateModel),
		KernelOperation(kernelOperationSummarizeModel),
		KernelOperation(kernelOperationDiagnoseModel),
	}
}

func moduleRoot() string {
	_, file, _, ok := runtime.Caller(0)
	if !ok {
		return filepath.FromSlash("bindings/go")
	}
	return filepath.Dir(file)
}

// KernelRequest mirrors the language-neutral kernel request envelope.
type KernelRequest struct {
	SchemaVersion string         `json:"schema_version"`
	Operation     string         `json:"operation"`
	ModelKey      *string        `json:"model_key"`
	Payload       map[string]any `json:"payload"`
	Metadata      map[string]any `json:"metadata"`
}

// KernelResponse mirrors the language-neutral kernel response envelope.
type KernelResponse struct {
	SchemaVersion string         `json:"schema_version"`
	Operation     string         `json:"operation"`
	ModelKey      string         `json:"model_key"`
	Payload       map[string]any `json:"payload"`
	Metadata      map[string]any `json:"metadata"`
	Error         *KernelError   `json:"error,omitempty"`
}

// KernelError mirrors the stable error envelope returned by the kernel bridge.
type KernelError struct {
	Code      string         `json:"code"`
	Message   string         `json:"message"`
	Operation string         `json:"operation,omitempty"`
	Details   map[string]any `json:"details"`
	Retryable bool           `json:"retryable"`
}
